File: src/upscale-plan.ts
/** Each fixed 2x request stays within the official 3 MP input envelope (checked 2026-09-21). */
export function planUpscale(width: number, height: number, scale: 2 | 4 | "max") {
  if (![width, height].every(v => Number.isFinite(v) && v > 0)) throw new Error("Invalid image dimensions");
  if (scale !== 2 && scale !== 4 && scale !== "max") throw new Error("Invalid upscale mode");
  const maxInputPixels = 3145728;
  const maxSide = 4096;
  let factor: number;
  if (scale === "max") {
    factor = Math.min(4, maxSide / width, maxSide / height, Math.sqrt(4 * maxInputPixels / (width * height)));
  } else {
    factor = Math.min(1, Math.sqrt(maxInputPixels / (width * height))) * scale;
  }
  const passes = scale === 4 || (scale === "max" && factor > 2) ? 2 : 1;
  const divisor = 2 ** passes;
  const inputWidth = Math.max(1, Math.floor(width * factor / divisor));
  const inputHeight = Math.max(1, Math.floor(height * factor / divisor));
  const outputWidth = inputWidth * divisor, outputHeight = inputHeight * divisor;
  const lastInputPixels = inputWidth * inputHeight * 4 ** (passes - 1);
  return {
    inputWidth, inputHeight, width: outputWidth, height: outputHeight, passes,
    resized: inputWidth !== width || inputHeight !== height,
    exceedsLimit: outputWidth > maxSide || outputHeight > maxSide || lastInputPixels > maxInputPixels,
  };
}

File: mobile/lib/images/upscale_plan.dart
import 'dart:math' as math;

// 0 is the persisted MAX choice; 2/4 retain their existing meaning.
({
  int inputWidth,
  int inputHeight,
  int width,
  int height,
  int passes,
  bool resized,
  bool exceedsLimit
}) planUpscale(int width, int height, int scale) {
  if (width <= 0 || height <= 0 || ![0, 2, 4].contains(scale)) {
    throw const FormatException('Invalid upscale size or mode');
  }
  const maxInputPixels = 3145728;
  final factor = scale == 0
      ? [
          4.0,
          4096 / width,
          4096 / height,
          math.sqrt(4 * maxInputPixels / (width * height))
        ].reduce(math.min)
      : math.min(1.0, math.sqrt(maxInputPixels / (width * height))) * scale;
  final passes = scale == 4 || (scale == 0 && factor > 2) ? 2 : 1;
  final divisor = 1 << passes;
  final iw = math.max(1, (width * factor / divisor).floor());
  final ih = math.max(1, (height * factor / divisor).floor());
  final ow = iw * divisor, oh = ih * divisor;
  return (
    inputWidth: iw,
    inputHeight: ih,
    width: ow,
    height: oh,
    passes: passes,
    resized: iw != width || ih != height,
    exceedsLimit: ow > 4096 ||
        oh > 4096 ||
        iw * ih * math.pow(4, passes - 1) > maxInputPixels
  );
}
